---
title: "raw_data_analysis"
output: html_document
date: "2022-09-21"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = '/Users/cgu3/Library/CloudStorage/OneDrive-InsideMDAnderson/data/missing-protein-project')
getwd()
```


```{r}
library(tidyverse)
# install.packages("devtools")
if (!require("protools", quietly = TRUE))
   devtools::install_github("https://github.com/FDUguchunhui/protools")
library('protools')
```


# Import data
```{r}
# import gastric primary cell nsaf data that filtered with Spectral count (SpC) >=2
gastric_nsaf <- as.matrix(read.csv('data/gastric-primary-cell/gastric_primary_cell_nsaf_2.csv', row.names = 1))
# # import gastric primary cell SpC data
gastric_count <- read_csv('data/gastric-primary-cell/gastric_primary_cell_count.csv', show_col_types = FALSE)
# rename the index column as "accession"
colnames(gastric_count)[1] <- 'accession'
# import information for mapping between accession, gene symbol, and ensembl ID
omics_metadata <- read_csv('data/support-data/omics_metadata.csv', show_col_types = FALSE)
# load missing protein information, it will load just one object missing_protein_df
load('cache/missing_protein.rda')
```
# Create regular protein set and missing protein set
create a character vector of missing protein accession only use P2-4 proteins from nextprot
the regular protein is not the just a complementary of missing protein, need to filter P1-5 proteins
```{r}
# missing_proteins is a character vector of MP gene symbols
# similar P2_5_proteins are protein gene symbols under P2-5 category in nextprot
# P2_5_proteins will be used for negative filtering to get regular proteins
missing_proteins <- missing_protein_df %>% filter(PE != 'Uncertain') %>% select(`gene name(s)`) %>% pull
P2_5_proteins <- missing_protein_df  %>% select(`gene name(s)`) %>% pull

# missing_proteins is a character vector of MP accession ID
# similar P2_5_proteins are protein accesion ID under P2-5 category in nextprot
missing_proteins_accession <- missing_protein_df %>% filter(PE != 'Uncertain') %>% select(Accession) %>% pull
P2_5_proteins_accession <- missing_protein_df %>% select(Accession) %>% pull
```





# check the number of PSM for each proteins for missing and regular proteins
```{r}
MP_counts <- gastric_count %>% filter(accession %in% missing_proteins_accession)
MP_counts_long <- MP_counts %>% pivot_longer(names_to = 'IPAS', values_to = 'count', cols = c(-accession)) %>% filter(count > 0)


RP_counts <-  gastric_count %>% filter(!(accession %in% P2_5_proteins_accession))
RP_counts_long <- RP_counts %>% pivot_longer(names_to = 'IPAS', values_to = 'count', cols = c(-accession)) %>% filter(count > 0)
```


```{r}
MP_counts_long %>% ggplot(aes(x=count)) + 
  geom_histogram()

MP_counts_long %>% ggplot(aes(x=count)) + 
  geom_histogram(breaks=seq(1, 10, 1)) +
   scale_x_continuous(breaks = seq(1, 10, 1), lim = c(1, 10)) +
  theme_classic()

RP_counts_long %>% ggplot(aes(x=count)) + 
  geom_histogram(breaks=seq(1, 300, 1))
RP_counts_long %>% ggplot(aes(x=count)) + 
  geom_histogram(breaks=seq(1, 10, 1)) +
   scale_x_continuous(breaks = seq(1, 10, 1), lim = c(1, 10)) +
  theme_classic()
```


```{r}

```







# create missing protein only use P2-4
# the regular protein is not the just a complementary of missing protein, need to filter P1-5 proteins
```{r}
# missing_protein_df <- missing_protein_df %>% filter(PE == 'Evidence at transcript level')
missing_proteins <- missing_protein_df %>% filter(PE != 'Uncertain') %>% select(`gene name(s)`) %>% pull
P2_5_proteins <- missing_protein_df  %>% select(`gene name(s)`) %>% pull

missing_proteins_accession <- missing_protein_df %>% filter(PE != 'Uncertain') %>% select(Accession) %>% pull
P2_5_proteins_accession <- missing_protein_df %>% select(Accession) %>% pull
```

# check the number of PSM for each proteins for missing and regular proteins
```{r}
# IP0981_1701 IP0982_1701 IP0993_1701 IP0995_1701 IP0999_1701 IP7100_1701 IP7103_1701 IP7105_1701 sample 1-8
colnames(gastric_count) <- c('accession', paste('Sample', 1:8, sep = ' '))
MP_counts <- gastric_count %>% filter(accession %in% missing_proteins_accession)
MP_counts_long <- MP_counts %>% pivot_longer(names_to = 'IPAS', values_to = 'count', cols = c(-accession)) %>% filter(count > 0)
MP_counts_long$type <- 'Missing'

RP_counts <-  gastric_count %>% filter(!(accession %in% P2_5_proteins_accession))
RP_counts_long <- RP_counts %>% pivot_longer(names_to = 'IPAS', values_to = 'count', cols = c(-accession)) %>% filter(count > 0)
RP_counts_long$type <- 'Regular'

all_counts_long <- rbind(MP_counts_long, RP_counts_long)
```


```{r}
RP_barplot_data <-  RP_counts_long %>% mutate(count_discrete=ifelse(count >= 50, '50+', count))
RP_barplot_data$count_discrete <- factor(RP_barplot_data$count_discrete, levels=c(as.character(1:49), '50+'))
RP_barplot_data %>% 
  ggplot(aes(x=count_discrete)) +
  geom_bar(position = 'identity', fill='blue') +
theme_bw() +
  xlab('Spectral counts') +
  ylab('Frequency') +
  ggtitle('Spectral count frequency of regular proteins') +
  scale_x_discrete(breaks=c('1', seq(5, 50, 5), '50+'))
  
  
```


```{r}
MP_barplot_data <-  MP_counts_long %>% mutate(count_discrete=ifelse(count >= 50, '50+', count))
MP_barplot_data$count_discrete <- factor(MP_barplot_data$count_discrete, levels=c(as.character(1:49), '50+'))
MP_barplot_data %>% 
  ggplot(aes(x=count_discrete)) +
  geom_bar(position = 'identity', fill='red') +
theme_bw() +
  xlab('Spectral counts') +
  ylab('Frequency') +
  ggtitle('Spectral count frequency of missing proteins') +
  scale_x_discrete(breaks=c('1', seq(5, 50, 5), '50+'), drop=FALSE)
  
  
  
```



```{r}
omics_metadata %>% filter(!(accession_id %in% P2_5_proteins_accession)) %>% summarise(avg_length=mean(protein_length, na.rm=TRUE))
omics_metadata %>% filter(accession_id %in% missing_proteins_accession) %>% summarise(avg_length=mean(protein_length, na.rm=TRUE))
```



# aggregate into gene-level
```{r}
# accession_gene_symbol <- data.frame(gene_symbol = accession_metadata$gene,row.names = accession_metadata$accession)
# rownames(gastric_nsaf) <- accession_gene_symbol[rownames(gastric_nsaf),]
gastric_nsaf_gene_level <- rownames_mapping(gastric_nsaf, omics_metadata, map_from = 'accession_id', map_to = 'gene_symbol')
# use sum or average? ?? sum seems to make more sense
temp <- as_tibble(gastric_nsaf_gene_level, rownames='symbol') %>%  filter(!is.na(symbol))  %>% group_by(symbol) %>% summarise_all(sum)
gastric_nsaf_gene_level <- as.matrix(temp[-1])
rownames(gastric_nsaf_gene_level) <- temp$symbol

head(gastric_nsaf_gene_level)
nrow(gastric_nsaf_gene_level)
```


```{r}
load('cache/count_TPM.rda')
count_TPM_gene_symbol <- rownames_mapping(count_TPM_mat, omics_metadata, map_from = 'ensembl_gene_id', map_to = 'gene_symbol')
count_TPM_gene_symbol <- count_TPM_gene_symbol[rowSums(count_TPM_gene_symbol > 0) > 0, ]

temp <- as_tibble(count_TPM_gene_symbol, rownames='symbol') %>% filter(!is.na(symbol)) %>% group_by(symbol) %>% summarise_all(sum)
count_TPM_gene_symbol <- as.matrix(temp[-1])
rownames(count_TPM_gene_symbol) <- temp$symbol
# change the column name the same as in spectral count matrix
colnames(count_TPM_gene_symbol) <- c('IP7103_1701', 'IP0972_1701', 'IP0995_1701', 'IP0982_1701', 'IP0999_1701', 'IP7100_1701', 'IP0993_1701', 'IP0981_1701', 'IP7105_1701')
count_TPM_gene_symbol <- count_TPM_gene_symbol[, colnames(gastric_nsaf_gene_level)]

head(count_TPM_gene_symbol)
nrow(count_TPM_gene_symbol)
write.csv(as.data.frame(count_TPM_gene_symbol), file='TPM_gastric_cancer_primary_cell.csv')
```

# rename to sample 1- 8
```{r}
# colnames(gastric_nsaf_gene_level) <- paste('Sample', 1:8, sep = ' ')
# colnames(count_TPM_gene_symbol) <- paste('Sample', 1:8, sep = ' ')
```


```{r}
count_TPM_gene_symbol <- count_TPM_gene_symbol[!is.na(rownames(count_TPM_gene_symbol)), ]
gastric_nsaf_gene_level <- gastric_nsaf_gene_level[!is.na(rownames(gastric_nsaf_gene_level)), ]
```

# check the overall covarage of RNA and protein for each other
```{r}
# percentage of gene that can be found to have protein product
mean(rownames(count_TPM_gene_symbol) %in% rownames(gastric_nsaf_gene_level))
# percentage of proteins that have transcription level evidence
mean(rownames(gastric_nsaf_gene_level) %in% rownames(count_TPM_gene_symbol))
```

# combine the RNA and protein data into long format
```{r}
A <- as_tibble(log(count_TPM_gene_symbol+1), rownames='accession') %>% pivot_longer(names_to = 'IPAS', values_to = 'TPM', cols = c(-accession))
B <- as_tibble(log(gastric_nsaf_gene_level+1), rownames='accession') %>% pivot_longer(names_to = 'IPAS', values_to = 'NSAF', cols = c(-accession))
combined_long <- full_join(A, B, by=c('accession', 'IPAS')) %>% replace(is.na(.), 0) %>% filter(!(TPM == 0 & NSAF == 0))
```


```{r}
#' @title calculate proportion of protein products supported with RNA products
#' @description It product the proportion of each condition based on different settings
#' @param tbl a tibble object, see details in "dplyr"
#' @param missing_protein whether to only include missing proteins
#' @param group_by_IPAS whether to stratified by IPAS samples
#' @param TPM_cutoff float, the cutoff used for RNA TPM value to consider a RNA product to be existing
#' @param NSAF_cutoff float, the cutoff used for RNA TPM value to consider a protein product to be existing
#' @param cases a vector of integer, used to decide whether cases to included for calculation
summarize_by_case <- function(tbl, proteins_subset, negate=FALSE, remove_IPAS=NULL, group_by_IPAS=FALSE, TPM_cutoff = 0, NSAF_cutoff = 0, cases = c(1, 2, 4)) {
  
  if(negate) {
    tbl <- tbl %>% filter(!(accession %in% proteins_subset))
  } else{
    tbl <- tbl %>% filter(accession %in% proteins_subset)
  }
  

  if (!is.null(remove_IPAS)) {
    tbl <- tbl %>% filter(!(IPAS %in% remove_IPAS))
  }
  
  tbl <- tbl %>% rowwise() %>%  
    mutate(case=if_else(TPM > TPM_cutoff & NSAF > NSAF_cutoff, 1,
                        if_else(TPM <= TPM_cutoff & NSAF > NSAF_cutoff, 2, 
                                if_else(TPM > TPM_cutoff & NSAF <= NSAF_cutoff, 3, 4)
                        )
    )
    ) %>% 
    filter(case %in% cases) %>% 
    group_by(IPAS, case) %>% 
    summarise(count=n(), ) %>% 
    ungroup(IPAS, case)
  
  if (group_by_IPAS) {
    tbl <- tbl %>% group_by(IPAS)
  }
  
  tbl <- tbl %>%
    group_by(case, .add=TRUE) %>% 
    summarise(count= sum(count)) %>%
    mutate(countT=sum(count)) %>% 
    mutate(per=paste0(round(100*count/countT,2),'%'))
  
  return(tbl)
}

```


# the proportion of PE2-5 in the missing protein dataset
```{r}
table(missing_protein_df$PE)
```


# check missing proteins that are constant detected in several samples
```{r}
# the total number of unique missing proteins
# freq <- combined_long %>% filter(accession %in% missing_proteins) %>% filter(NSAF > 0) %>% select(accession) %>% unique()

detected_MP_proteins <- combined_long %>%  filter(accession %in% missing_proteins) %>% filter(NSAF > 0  & TPM > 0) %>% arrange(accession) %>% group_by(accession) 
# %>% summarise(n=n()) %>% filter(n >= 2)
# hist(detected_MP_proteins$n)


# check the protein and RNA expression for the detected missing proteins
full_table_detected_MP <- combined_long %>% filter(accession %in% missing_proteins) %>% filter(NSAF > 0 & TPM > 0) %>% arrange(accession, IPAS) %>% filter(accession %in% detected_MP_proteins$accession)
full_table_detected_MP <- inner_join(full_table_detected_MP, detected_MP_proteins)

# 
left_join(full_table_detected_MP, MP_counts_long, by = c('accession', 'IPAS'))

# MP_count_mat <- as.matrix(MP_counts[-1])
# rownames(MP_count_mat) <- MP_counts$accession
# MP_count_mat_gene_level <- rownames_mapping(MP_count_mat, omics_metadata, map_from = 'accession_id', map_to = 'gene_symbol')

write_csv(full_table_detected_MP, 'output/protein list/detected_104_MP_products_with_RNA.csv')

```

```{r}
cor(full_table_detected_MP$NSAF, full_table_detected_MP$TPM, method = 'spearman')
```

```{r}
# library(ggrepel)
# 
# MPs_scatterplot <- ggplot(full_table_detected_MP, aes(x=TPM, y=NSAF,  label=accession)) +
#   geom_point()
# MPs_scatterplot +
#   # geom_text_repel(data = subset(full_table_detected_MP, NSAF > 2 & TPM > 2)) +
#   geom_label_repel(data = subset(full_table_detected_MP, n > 2),
#                    aes(color=accession),
#                    ox.padding   = 0.35, 
#                   point.padding = 0.5,
#                   segment.color = 'grey50',
#                   direction     = "x",
#                   max.overlaps = 50, 
#                   force=30) +
#   
#     theme_classic() +
#   theme(legend.position="none")
```





```{r}
mapping_to_entrez <- function(gene_names) {
  entrez_id <- AnnotationDbi::select(org.Hs.eg.db::org.Hs.eg.db, gene_names, keytype = "SYMBOL", columns=c("ENTREZID", "UNIPROT"))$ENTREZID
  return(entrez_id)
}
```


```{r eval=FALSE, include=FALSE}
pathways_analysis_MP <- limma::goana(mapping_to_entrez(detected_MP_proteins$accession), plot=TRUE)

expressed_regular_proteins <- combined_long %>% filter(!(IPAS %in% c('IP0981_1701', 'IP7103_1701'))) %>%  filter(!(accession %in% P2_5_proteins)) %>% filter(NSAF > 0) %>% arrange(accession) %>% group_by(accession) %>% summarise(n=n()) %>% filter(n >= 3)
pathways_analysis_RP <- limma::goana(mapping_to_entrez(expressed_regular_proteins$accession), plot=TRUE)


all_expressed_proteins <- union(expressed_regular_proteins$accession, detected_MP_proteins$accession)
write_csv(tibble(accession=all_expressed_proteins), 'data/expressed-protein-reference-set.csv')
pathways_analysis_all <- limma::goana(mapping_to_entrez(all_expressed_proteins), plot=TRUE)
```

```{r eval=FALSE, include=FALSE}
(topGOs_MP <- limma::topGO(pathways_analysis_MP, ontology = 'BP', number=20))
```

```{r eval=FALSE, include=FALSE}
(topGOs_RP <- limma::topGO(pathways_analysis_RP, ontology = 'BP', number = 20))
```


```{r eval=FALSE, include=FALSE}
intersect(topGOs_MP$Term, topGOs_RP$Term)
```



```{r eval=FALSE, include=FALSE}
topGOs_all <- (limma::topGO(pathways_analysis_all, ontology = 'BP', number=20))
```


```{r eval=FALSE, include=FALSE}
intersect(topGOs_RP$Term, topGOs_all$Term)
```
```{r eval=FALSE, include=FALSE}
topGOs_all$Term[!topGOs_all$Term %in% topGOs_RP$Term]
```

```{r eval=FALSE, include=FALSE}
topGOs_MP$Term[!topGOs_MP$Term %in% topGOs_RP$Term]
```




```{r}
combined_long %>% filter(accession == 'CTAGE1')
```





############################ association analysis#####################
# Table 1
```{r}
# the coverage rate of RNA product in protein product for missing protein
(a <- summarize_by_case(combined_long, proteins_subset=missing_proteins, group_by_IPAS = TRUE, cases = c(1, 2)))
write_csv(a, file='output/sample_MP_stats.csv')

## overall without removing the outlier
(b <- summarize_by_case(combined_long, proteins_subset=missing_proteins, group_by_IPAS = FALSE, cases = c(1, 2)))

## overall with removing the outlier
(c <- summarize_by_case(combined_long, proteins_subset=missing_proteins, group_by_IPAS = FALSE, cases = c(1, 2), remove_IPAS = c('IP0981_1701', 'IP7103_1701')))

# the coverage rate of protein product in RNA product for missing protein
summarize_by_case(combined_long, proteins_subset=missing_proteins, group_by_IPAS = TRUE, cases = c(1, 3))

## overall without removing the outlier
summarize_by_case(combined_long, proteins_subset=missing_proteins, group_by_IPAS = FALSE, cases = c(1, 3))

## overall with removing the outlier
summarize_by_case(combined_long, proteins_subset=missing_proteins, group_by_IPAS = FALSE,  cases = c(1, 3), remove_IPAS = c('IP0981_1701', 'IP7103_1701'))
```



histogram
```{r}
b$IPAS = 'Combined'
c$IPAS = 'Combined w.o. outlier'
dat_hist <- rbind(a, b, c)


dat_hist$case <- factor(dat_hist$case, levels = c('1', '2'), labels = c('RNA (+)', 'RNA (-)'))
dat_hist$case <- relevel(dat_hist$case, 'RNA (+)')

dat_hist$IPAS <- factor(dat_hist$IPAS, levels = c("IP0981_1701", "IP0982_1701", "IP0993_1701", "IP0995_1701", "IP0999_1701",
                                                  "IP7100_1701", "IP7103_1701", "IP7105_1701", "Combined w.o. outlier",  "Combined"))


# ggplot(data=dat_hist, aes(x=IPAS, y=count, group=case, fill=case)) +
#   labs(title="Coverage rate for prodein product with RNA product in missing proteins", subtitle = '',
#         x = "IPAS", y = 'count') + 
#   geom_bar( stat = 'identity',  position=position_dodge()) + 
#   geom_text(aes(label=per), color="black",size=3, vjust=-0.5, position = position_dodge(1)) +
#   theme(
#         axis.text.x = element_text(angle = 45, hjust = 1))

```




# Table 2 non-missing proteins

```{r eval=FALSE, include=FALSE}
# the coverage rate of RNA product in protein product for non-missing protein
(a <- summarize_by_case(combined_long, proteins_subset = P2_5_proteins, negate = TRUE, group_by_IPAS = TRUE, cases = c(1, 2)))

## overall without removing the outlier
(b <- summarize_by_case(combined_long, proteins_subset = P2_5_proteins, negate = TRUE, group_by_IPAS = FALSE, cases = c(1, 2)))

## overall with removing the outlier
(c <- summarize_by_case(combined_long, proteins_subset = P2_5_proteins, negate = TRUE, group_by_IPAS = FALSE, cases = c(1, 2), remove_IPAS = c('IP0981_1701', 'IP7103_1701')))

# the coverage rate of protein product in RNA product for non-missing protein
summarize_by_case(combined_long, proteins_subset = P2_5_proteins, negate = TRUE, group_by_IPAS = TRUE, cases = c(1, 3))

## overall without removing the outlier
summarize_by_case(combined_long, proteins_subset = P2_5_proteins, negate = TRUE, group_by_IPAS = FALSE, cases = c(1, 3))

## overall with removing the outlier
summarize_by_case(combined_long, proteins_subset = P2_5_proteins, negate = TRUE, group_by_IPAS = FALSE, cases = c(1, 3), remove_IPAS = c('IP0981_1701', 'IP7103_1701'))
```

## barplot for non-missing proteins
```{r}
b$IPAS = 'Combined'
c$IPAS = 'Combined2'
dat_hist <- rbind(a, b, c)


dat_hist$case <- factor(dat_hist$case, levels = c('1', '2'), labels = c('RNA (+)', 'RNA (-)'))
dat_hist$case <- relevel(dat_hist$case, 'RNA (+)')

dat_hist$IPAS <- factor(dat_hist$IPAS, levels = c("IP0981_1701", "IP0982_1701", "IP0993_1701", "IP0995_1701", "IP0999_1701",
                                                  "IP7100_1701", "IP7103_1701", "IP7105_1701", "Combined",  "Combined2"))


# ggplot(data=dat_hist, aes(x=IPAS, y=count, group=case, fill=case)) +
#   labs(title="Coverage rate for prodein product with RNA product in non-missing proteins", subtitle = '',
#         x = "IPAS", y = 'count') + 
#   geom_bar( stat = 'identity',  position=position_dodge()) + 
#   geom_text(aes(label=per), color="black",size=3, vjust=-0.5, position = position_dodge(1)) +
#   theme(
#         axis.text.x = element_text(angle = 45, hjust = 1))

```

<!-- # Table 3 all proteins -->
<!-- ```{r eval=FALSE, include=FALSE} -->
<!-- # the coverage rate of RNA product in protein product for missing protein -->
<!-- summarize_by_case(combined_long, protein_subset = 'all', group_by_IPAS = TRUE, missing_proteins = missing_protein_df$`gene name(s)`, cases = c(1, 2), TPM_cutoff = 0, NSAF_cutoff = 0) -->

<!-- ## overall without removing the outlier -->
<!-- summarize_by_case(combined_long, protein_subset = 'all', group_by_IPAS = FALSE, missing_proteins = missing_protein_df$`gene name(s)`, cases = c(1, 2), TPM_cutoff = 0, NSAF_cutoff = 0) -->

<!-- ## overall with removing the outlier -->
<!-- summarize_by_case(combined_long, protein_subset = 'all', group_by_IPAS = FALSE, missing_proteins = missing_protein_df$`gene name(s)`, cases = c(1, 2), TPM_cutoff = 0, NSAF_cutoff = 0, remove_IPAS = c('IP7103_1701')) -->

<!-- # the coverage rate of protein product in RNA product for missing protein -->
<!-- summarize_by_case(combined_long, protein_subset = 'all', group_by_IPAS = TRUE, missing_proteins = missing_protein_df$`gene name(s)`, cases = c(1, 3), TPM_cutoff = 0, NSAF_cutoff = 0) -->

<!-- ## overall without removing the outlier -->
<!-- summarize_by_case(combined_long, protein_subset = 'all', group_by_IPAS = FALSE, missing_proteins = missing_protein_df$`gene name(s)`, cases = c(1, 3), TPM_cutoff = 0, NSAF_cutoff = 0) -->

<!-- ## overall with removing the outlier -->
<!-- summarize_by_case(combined_long, protein_subset = 'all', group_by_IPAS = FALSE, missing_proteins = missing_protein_df$`gene name(s)`, cases = c(1, 3), TPM_cutoff = 0, NSAF_cutoff = 0, remove_IPAS = c('IP7103_1701')) -->
<!-- ``` -->

<!-- ! NA %in% NA is treated as 0 in R so need to remove it before use to get the intersection -->
<!-- of gene symbols between two data -->
<!-- ```{r} -->
<!-- both <- na.omit(intersect(rownames(count_TPM_gene_symbol), rownames(gastric_nsaf_gene_level))) -->
<!-- length(both) -->
<!-- ``` -->


<!-- # ? how about keep the largest? and fill 0 in the other part? -->
<!-- ```{r} -->
<!-- count_TPM_both <- count_TPM_gene_symbol[rownames(count_TPM_gene_symbol) %in% both,] -->
<!-- count_nsaf_both <- gastric_nsaf_gene_level[rownames(gastric_nsaf_gene_level) %in% both,] -->
<!-- ``` -->



<!-- reorder row to make the two matrix have the same row order -->
<!-- ```{r} -->
<!-- count_nsaf_both <- count_nsaf_both[both,] -->
<!-- count_TPM_both <- count_TPM_both[both, ] -->
<!-- head(count_nsaf_both) -->
<!-- head(count_TPM_both) -->
<!-- ``` -->

<!-- # the all-gene heatmap -->
<!-- ```{r eval=FALSE, include=FALSE} -->
<!-- pheatmap::pheatmap(log(count_TPM_both+1), -->
<!--                    cluster_rows = TRUE, -->
<!--                    cluster_cols = FALSE, -->
<!--                    show_rownames = FALSE -->
<!--                    ) -->
<!-- ``` -->



<!-- ```{r eval=FALSE, include=FALSE} -->
<!-- pheatmap::pheatmap(log(count_nsaf_both+1), -->
<!--                    cluster_rows = TRUE, -->
<!--                    cluster_cols = FALSE, -->
<!--                    show_rownames = FALSE -->
<!--                    ) -->
<!-- ``` -->



# combine the information
```{r}
A <- as_tibble(log(count_TPM_gene_symbol+1), rownames='gene_symbol') %>% pivot_longer(names_to = 'IPAS', values_to = 'TPM', cols = c(-gene_symbol))
B <- as_tibble(log(gastric_nsaf_gene_level+1), rownames='gene_symbol') %>% pivot_longer(names_to = 'IPAS', values_to = 'NSAF', cols = c(-gene_symbol))
glmm_dat_all <- full_join(A, B, by=c('gene_symbol', 'IPAS')) %>% replace(is.na(.), 0) %>% filter(!(TPM == 0 & NSAF == 0))
```

```{r}
plt <- ggplot(data      = glmm_dat_all,
       aes(x     = TPM,
           y     = NSAF,
           col   = IPAS,
           group = IPAS))+ #to add the colours for different classes
  geom_point(size     = 1.2,
             alpha    = .8,
             position = "jitter")+ #to add some random noise for plotting purposes
  theme_minimal() +
  scale_color_discrete(labels=paste('sample', 1:8)) +
 # to add regression line
  geom_vline(xintercept = 0.01, lty=2) + geom_hline(yintercept =0.1, lty=2) +
  labs(title    = "Normalized spectral abundance factor (NASF) vs. Transcripts per million (TPM)",
       subtitle = "missing proteins on gene level in gastric cancer primary cell samples",
       xlab = 'Log2 (TPM + 1)',
       ylab = 'Log2 (NSAF + 1)',
  )
    
  
  
dens1 <- ggplot(glmm_dat_all, aes(x = TPM, fill = IPAS)) + 
  geom_density(alpha = 0.4) + 
  theme_void() + 
  theme(legend.position = "none")

dens2 <- ggplot(glmm_dat_all, aes(x = NSAF, fill = IPAS)) + 
  geom_density(alpha = 0.4) + 
  theme_void() + 
  theme(legend.position = "none") + 
  coord_flip()


dens1 + patchwork::plot_spacer() + plt + dens2 +  patchwork::plot_layout(ncol = 2, nrow = 2, widths = c(4, 1), heights = c(1, 4))
```


# missing proteins
check how many missing proteins is expressed in the 8 gastric sample in the protein data and in RNA-seq data  
```{r}
mp_in_gene_symbol <- missing_protein_df$`gene name(s)`
sum(rownames(gastric_nsaf_gene_level) %in% missing_proteins)
# sum(both %in% mp_in_gene_symbol)
```


There two many gene, so let's only keep missing proteins
```{r}
count_nsaf_missing <-  gastric_nsaf_gene_level[rownames(gastric_nsaf_gene_level) %in% missing_proteins, ]
count_TPM_missing <- count_TPM_gene_symbol[rownames(count_TPM_gene_symbol) %in% missing_proteins, ]
count_nsaf_non_missing <-  gastric_nsaf_gene_level[!(rownames(gastric_nsaf_gene_level) %in% P2_5_proteins), ]
count_TPM_non_missing <- count_TPM_gene_symbol[!(rownames(count_TPM_gene_symbol) %in% P2_5_proteins), ]
# unique_gene_cover_rate <- mean(rownames(count_nsaf_missing) %in% rownames(count_TPM_missing))
# paste('The number of uniquely detected gene of missing proteins in MS count is: ', nrow(count_dat_missing), 'coverage rate is:', unique_gene_cover_rate, sep = ' ')
```

<!-- ```{r} -->
<!-- pheatmap::pheatmap(log(count_TPM_missing+1), -->
<!--                    cluster_rows = FALSE, -->
<!--                    cluster_cols = FALSE, -->
<!--                    show_rownames = FALSE -->
<!--                    ) -->
<!-- ``` -->

<!-- ```{r} -->
<!-- pheatmap::pheatmap(log(count_nsaf_missing+1), -->
<!--                    cluster_rows = FALSE, -->
<!--                    cluster_cols = FALSE, -->
<!--                    show_rownames = FALSE -->
<!--                    ) -->
<!-- ``` -->

# check paired correlation
```{r}
A <- as_tibble(log(count_TPM_missing+1), rownames='accession') %>% pivot_longer(names_to = 'IPAS', values_to = 'TPM', cols = c(-accession))
B <- as_tibble(log(count_nsaf_missing+1), rownames='accession') %>% pivot_longer(names_to = 'IPAS', values_to = 'NSAF', cols = c(-accession))
glmm_dat_missing <- full_join(A, B, by=c('accession', 'IPAS')) %>% replace(is.na(.), 0) %>% filter(!(TPM == 0  & NSAF == 0))
glmm_dat_missing_filtered <- glmm_dat_missing %>% filter(TPM > 0 & NSAF > 0)

A <- as_tibble(log(count_TPM_non_missing+1), rownames='accession') %>% pivot_longer(names_to = 'IPAS', values_to = 'TPM', cols = c(-accession))
B <- as_tibble(log(count_nsaf_non_missing+1), rownames='accession') %>% pivot_longer(names_to = 'IPAS', values_to = 'NSAF', cols = c(-accession))
glmm_dat_non_missing <- full_join(A, B, by=c('accession', 'IPAS')) %>% replace(is.na(.), 0) %>% filter(!(TPM == 0  & NSAF == 0))
glmm_dat_non_missing_filtered <- glmm_dat_non_missing %>% filter(TPM > 0 & NSAF > 0)
```



<!-- # Generalization linear mixed model -->
<!-- ```{r} -->
<!-- library(lme4) -->
<!-- glmm_model <- lmer(NSAF ~ TPM + (TPM | IPAS), data = glmm_dat) -->
<!-- summary(glmm_model) -->
<!-- ``` -->

<!-- ```{r} -->
<!-- ggplot(data      = glmm_dat, -->
<!--        aes(x     = TPM, -->
<!--            y     = NSAF, -->
<!--            col   = IPAS, -->
<!--            group = IPAS))+ #to add the colours for different classes -->
<!--   geom_point(size     = 1.2, -->
<!--              alpha    = .8, -->
<!--              position = "jitter")+ #to add some random noise for plotting purposes -->
<!--   theme_minimal() + -->
<!--   geom_smooth(method = lm, -->
<!--               se     = FALSE, -->
<!--               size   = .5,  -->
<!--               alpha  = .8)+ # to add regression line -->
<!--   labs(title    = "Normalized spectral abundance factor (NASF) vs. Transcripts per million (TPM)", -->
<!--        subtitle = "missing proteins on gene level in gastric cancer primary cell samples", -->
<!--        xlab = 'Log2(TPM) + 1', -->
<!--        ylab = 'Log2(NSAF) + 1', -->
<!--   ) -->
<!-- ``` -->


```{r}
model.lm <- lm(NSAF ~ TPM, data = glmm_dat_missing_filtered)
summary(model.lm)
```



<!-- ```{r} -->

<!-- plt <- ggplot(data = glmm_dat_missing, -->
<!--        aes(x     = TPM, -->
<!--            y     = NSAF, -->
<!--            col   = IPAS) -->
<!--            )+ #to add the colours for different classes -->
<!--   geom_point(size     = 1.2, -->
<!--              alpha    = .8, -->
<!--              position = "jitter") + #to add some random noise for plotting purposes -->
<!--   geom_vline(xintercept=0.1, linetype=2, color='red', size=0.5) +  -->
<!--   geom_hline(yintercept=0.1, linetype=2, color='red', size=0.5) +  -->
<!--   theme_minimal() + -->
<!--   labs(title    = "Normalized spectral abundance factor (NASF) vs. Transcripts per million (TPM)", -->
<!--        subtitle = "missing proteins on gene level in gastric cancer primary cell samples", -->
<!--        xlab = 'Log2(TPM) + 1', -->
<!--        ylab = 'Log2(NSAF) + 1') -->

<!-- dens1 <- ggplot(glmm_dat_missing, aes(x = TPM, fill = IPAS)) +  -->
<!--   geom_density(alpha = 0.4) +  -->
<!--   theme_void() +  -->
<!--   theme(legend.position = "none") -->

<!-- dens2 <- ggplot(glmm_dat_missing, aes(x = NSAF, fill = IPAS)) +  -->
<!--   geom_density(alpha = 0.4) +  -->
<!--   theme_void() +  -->
<!--   theme(legend.position = "none") +  -->
<!--   coord_flip() -->


<!-- dens1 + patchwork::plot_spacer() + plt + dens2 +  patchwork::plot_layout(ncol = 2, nrow = 2, widths = c(4, 1), heights = c(1, 4)) -->
<!-- ``` -->



# plot with spearman correlation for missing proteins
```{r}
par(mfrow=c(2, 2))
ggpubr::ggscatter(glmm_dat_missing, x = "TPM", y = "NSAF",
   color = 'IPAS', size = 1.5, # Points color, shape and size
   title='Plot of NSAF vs. TPM for missing proteins',
   xlab = 'Log2 (TPM + 1)',
   ylab = 'Log2 (NSAF + 1)',
   ) + 
  scale_color_discrete(labels=paste('sample', 1:8)) +
  scale_y_continuous(name='Log2 (NSAF + 1)', breaks=c(0, 2, 4, 6, 8, 10), limits = c(0, 10)) +
  # geom_vline(xintercept=0.1, linetype=2, color='red', size=0.5) +
  geom_hline(yintercept=2, linetype=2, color='red', size=0.5) +
  theme(legend.text = element_text(size=10)) + 
  labs(color='Sample')


# plot with spearman correlation for missing proteins
plt <- ggpubr::ggscatter(glmm_dat_missing_filtered, x = "TPM", y = "NSAF",
   color = 'IPAS', size = 1.5, # Points color, shape and size
   add = "reg.line",  # Add regressin line
   add.params = list(color = "blue", fill = "lightgray"), # Customize reg. line
   conf.int = TRUE, # Add confidence interval
   cor.coef = TRUE, # Add correlation coefficient. see ?stat_cor
   cor.coeff.args = list(method = "spearman", cor.coef.name = "rho", label.x.npc=0.8, label.sep = "\n",
                         label.y.npc = 0.2),
   title='Plot of NSAF vs. TPM for missing proteins with NSAF > 0 and TPM > 0',
   # subtitle='missing proteins on gene level in gastric cancer primary cell sample',
   xlab = 'Log2 (TPM + 1)',
   ylab = 'Log2 (NSAF + 1)',
   ) +
   scale_fill_discrete(labels=paste('sample', 1:8)) +
   scale_y_continuous(name='Log2 (NSAF + 1)', breaks=c(0, 2, 4, 6, 8, 10), limits = c(0, 8)) +
   theme(legend.text = element_text(size=10)) +
   labs(color='Sample')

dens1 <- ggplot(glmm_dat_missing_filtered, aes(x = TPM, fill = IPAS)) + 
  geom_density(alpha = 0.4) + 
  theme_void() + 
  theme(legend.position = "none")

dens2 <- ggplot(glmm_dat_missing_filtered, aes(x = NSAF, fill = IPAS)) + 
  geom_density(alpha = 0.4) + 
  theme_void() + 
  theme(legend.position = "none") + 
  coord_flip()

dens1 + patchwork::plot_spacer() + plt + dens2 +  patchwork::plot_layout(ncol = 2, nrow = 2, widths = c(4, 1), heights = c(1, 4))


# plot with spearman correlation for non-missing proteins
ggpubr::ggscatter(glmm_dat_non_missing, x = "TPM", y = "NSAF",
   color = 'IPAS', size = 1.5, # Points color, shape and size
   title='Plot of NSAF vs. TPM for regular proteins',
   xlab = 'Log2 (TPM + 1)',
   ylab = 'Log2 (NSAF + 1)') +
  scale_color_discrete(labels=paste('sample', 1:8)) +
   scale_x_continuous(name='Log2 (TPM + 1)', breaks=seq(0, 16, 2), limits = c(0, 16)) +
   scale_y_continuous(name='Log2 (NSAF + 1)', breaks=seq(0, 14, 2), limits = c(0, 14)) +
   theme(legend.text = element_text(size=10)) + 
  # geom_vline(xintercept=0.1, linetype=2, color='red', size=0.5) +
  geom_hline(yintercept=2, linetype=2, color='red', size=0.5) +
  labs(color='Sample')

plt <- ggpubr::ggscatter(glmm_dat_non_missing_filtered, x = "TPM", y = "NSAF",
   color = 'IPAS', size = 1.5, # Points color, shape and size
   add = "reg.line",  # Add regressin line
   add.params = list(color = "blue", fill = "lightgray"), # Customize reg. line
   conf.int = TRUE, # Add confidence interval
   cor.coef = TRUE, # Add correlation coefficient. see ?stat_cor
   cor.coeff.args = list(method = "spearman", cor.coef.name = "rho",  label.x.npc=0.85, label.sep = "\n",
                         label.y.npc = 0.2),
   title='Plot of NSAF vs. TPM for regular proteins with NSAF > 0 and TPM > 0',
   # subtitle='missing proteins on gene level in gastric cancer primary cell sample',
   xlab = 'Log2 (TPM + 1)',
   ylab = 'Log2 (NSAF + 1)') +
  scale_color_discrete(labels=paste('sample', 1:8)) +
   scale_x_continuous(name='Log2 (TPM + 1)', breaks=seq(0, 16, 2), limits = c(0, 16)) +
   scale_y_continuous(name='Log2 (NSAF + 1)', breaks=seq(0, 14, 2), limits = c(0, 14)) +
    theme(legend.text = element_text(size=10)) +
  labs(color='Sample')

dens1 <- ggplot(glmm_dat_non_missing_filtered, aes(x = TPM, fill = IPAS)) + 
  geom_density(alpha = 0.4) + 
  theme_void() + 
  theme(legend.position = "none")

dens2 <- ggplot(glmm_dat_non_missing_filtered, aes(x = NSAF, fill = IPAS)) + 
  geom_density(alpha = 0.4) + 
  theme_void() + 
  theme(legend.position = "none") + 
  coord_flip()

dens1 + patchwork::plot_spacer() + plt + dens2 +  patchwork::plot_layout(ncol = 2, nrow = 2, widths = c(4, 1), heights = c(1, 4))
```



```{r}
# cor.mtest <- function(mat, conf.level = 0.95){
#   mat <- as.matrix(mat)
#     n <- ncol(mat)
#     p.mat <- lowCI.mat <- uppCI.mat <- matrix(NA, n, n)
#     diag(p.mat) <- 0
#     diag(lowCI.mat) <- diag(uppCI.mat) <- 1
#     for(i in 1:(n-1)){
#         for(j in (i+1):n){
#             tmp <- cor.test(mat[,i], mat[,j], conf.level = conf.level)
#             p.mat[i,j] <- p.mat[j,i] <- tmp$p.value
#             lowCI.mat[i,j] <- lowCI.mat[j,i] <- tmp$conf.int[1]
#             uppCI.mat[i,j] <- uppCI.mat[j,i] <- tmp$conf.int[2]
#         }
#     }
#     return(list(p.mat, lowCI.mat, uppCI.mat))
# }
# 
# res1 <- cor.mtest(mtcars,0.95)

# library(corrplot)
# corrplot(cor(count_nsaf_missing, count_TPM_missing, method = 'pearson'))
```

# Heatmap for missing protein
```{r}
# detected_MP_proteins <- combined_long %>% filter(!(IPAS %in% c('IP0981_1701', 'IP7103_1701'))) %>%  filter(accession %in% missing_proteins) %>% filter(NSAF > 0  & TPM > 0) %>% arrange(accession) %>% group_by(accession) %>% summarise(n=n()) %>% filter(n>= 2)  %>%  select(accession) %>% pull() %>% unique()

# the number should be 85 if using two-SpC-rule
detected_MP_proteins <- combined_long  %>%  filter(accession %in% missing_proteins) %>% filter(NSAF > 0  & TPM > 0) %>% arrange(accession) %>% group_by(accession) %>% summarise(n=n())  %>%  select(accession) %>% pull() %>% unique()
```

```{r}
# double check clustering
library(pheatmap)
heatmap_dat <- log(gastric_nsaf_gene_level[detected_MP_proteins, ] + 1)
colnames(heatmap_dat) =paste('sample', 1:8)
pheatmap(heatmap_dat, 
         show_rownames = T,
         angle_col=45,
         cluster_cols = T,
         cluster_rows = T)
```


```{r}
heatmap_dat_TPM <- log(count_TPM_gene_symbol[detected_MP_proteins, ] + 1)
colnames(heatmap_dat_TPM) =paste('sample', 1:8)
pheatmap(heatmap_dat_TPM, 
         show_rownames = T,
         angle_col=45,
         cluster_cols = T,
         cluster_rows = T)
```


## correlation plot for pairwise correlation of missing protein expression
```{r}
library(corrplot)
corr <- cor(heatmap_dat, method = "spearman")
corrplot.mixed(corr, lower = 'number', upper='ellipse', tl.cex=0.5,
               na.label.col='black',
               mar=c(0,0,2,0))
```


```{r}
corr <- cor(heatmap_dat_TPM, method = "spearman")
corrplot.mixed(corr, lower = 'number', upper='ellipse', tl.cex=0.6, na.label.col='black')
```

```{r}
corr <- cor(t(heatmap_dat), t(heatmap_dat_TPM), method = "spearman")
corr_diag <- diag(corr)
pheatmap(corr)
```



```{r}
# corr <- cor(count_nsaf_missing, count_TPM_missing)
# mean(diag(corr) > 0.2, na.rm = T)
# library(corrplot)
# corrplot(corr, method = 'color', na.label.col='black')
```



```{r}
detected_MP_proteins
```


```{r}
missing_protein_df %>% filter(`gene name(s)` %in% detected_MP_proteins) %>% select(PE) %>% pull() %>% table()
```



# Additional results

## Are the detected features gastric specific
```{r}
# misspelling in the original file name 
primary_cell_nsaf <- read.csv('data/nasf/nasf_primary_cell_2.csv', row.names = 1)
primary_cell_nsaf <- as.matrix(primary_cell_nsaf)
# read annotation file
IPAS_annotation <- read_csv('data/support-data/IPAS_annotation.csv')

# only keep missing protein
# primary_cell_nsaf <- primary_cell_nsaf[rownames(primary_cell_nsaf) %in% missing_proteins_accession,]

primary_cell_nsaf_gene_symbol <- rownames_mapping(primary_cell_nsaf, omics_metadata, map_from = 'accession_id', map_to = 'gene_symbol')
# use sum or average? ?? sum seems to make more sense
temp <- as_tibble(primary_cell_nsaf_gene_symbol, rownames='symbol') %>%  filter(!is.na(symbol))  %>% group_by(symbol) %>% summarise_all(sum)
primary_cell_nsaf_gene_symbol <- as.matrix(temp[-1])
rownames(primary_cell_nsaf_gene_symbol) <- temp$symbol
```
```{r}
primary_cell_nsaf['A0A087WTP8', 'IP0993_1701']
gastric_nsaf['A0A087WTP8', 'IP0993_1701']
```


```{r}
primary_cell_nsaf_gene_symbol['AGAP4',  'IP0993_1701']
gastric_nsaf_gene_level['AGAP4', 'IP0993_1701']
```


```{r}
primary_cell_nasf_gene_symbol <- primary_cell_nsaf_gene_symbol[!is.na(rownames(primary_cell_nsaf_gene_symbol)), ]
```



```{r}
IPAS_gastric <- IPAS_annotation %>% filter(disease=='Gastric') %>% select(ipas) %>% pull()
```


```{r}
create_annoataion <- function(data, annotation_table) {
  annotation_tbl <- annotation_table %>% select(ipas, disease, subtype) %>% filter(ipas %in% colnames(data))
  annotation_df <- data.frame(disease=annotation_tbl$disease, subtype=annotation_tbl$subtype)
  rownames(annotation_df) <- annotation_tbl$ipas
  return(annotation_df)
}

primary_cell_annotation <- create_annoataion(primary_cell_nasf_gene_symbol, annotation_table = IPAS_annotation)
primary_cell_annotation$type <- paste(primary_cell_annotation$disease, primary_cell_annotation$subtype)
primary_cell_annotation <- primary_cell_annotation['type']
colnames(primary_cell_annotation) <- 'disease'
table(primary_cell_annotation$disease)
```



```{r}
extract_sample_id <- function(x, disease) {
  rownames(x$annotation[x$annotation$disease %in% disease, ,drop=FALSE])
}

# diseases_keep <- c('Colon', 'Glioma', 'Leukemia', 'Lymphoblast', 'Melanoma', 'Neuroblastoma', 'Prostate', 'Ovarian')
diseases_keep <- c("Leukemia ALL", "Leukemia AML", "Leukemia MDS", "Ovarian Ascites", "Gastric Ascites")


SpC_list_primary_cell <- SpC_List(primary_cell_nasf_gene_symbol, annotation = primary_cell_annotation, proteins_filter=missing_proteins)

replicates_for_keep <- extract_sample_id(SpC_list_primary_cell, diseases_keep)
```

```{r}
# dat_primary_cell <- SpC_List(primary_cell_nasf_gene_symbol, annotation = primary_cell_annotation, replicates_keep = replicates_for_keep)
# dat_primary_cell$matrix <- dat_primary_cell$matrix[rowSums(dat_primary_cell$matrix) > 0, ]
```


# create SpC_List object for each disease 
```{r}
IPAS <- mapply(FUN=extract_sample_id, disease=c("Leukemia ALL", "Leukemia AML", "Leukemia MDS", "Ovarian Ascites", "Gastric Ascites"), MoreArgs = list(x=SpC_list_primary_cell))
SpC_lists <- mapply(FUN=SpC_List, replicates_keep=IPAS, MoreArgs = list(df=primary_cell_nasf_gene_symbol, annotation=primary_cell_annotation, proteins_filter=missing_proteins), SIMPLIFY = FALSE)
```


```{r}
prefilter <- function(x) {
  # N <- ncol(x$matrix)/2
  N <- 2
  keep <- rowSums(x$matrix > 0) >= N
  return(rownames(x$matrix[keep,]))
}
```


```{r}
MPs <- lapply(SpC_lists, prefilter)
sapply(MPs, length)
```


```{r}
plot_venn_diagram(MPs)
```


```{r}
VennDiagram::get.venn.partitions(MPs)
```
```{r}
data.frame(data = names(MPs), number = matrix(MPs)) %>%
  tidyr::unnest(cols = c(number)) %>%
  dplyr::distinct(number, .keep_all = TRUE)
```


```{r}
'RFPL2' %in% detected_MP_proteins
'TBC1D28' %in% detected_MP_proteins
'ZNF582' %in% detected_MP_proteins
```



# ```{r}
# (all_nasf_gene_symbol[, IPAS_breast] > 0)
# ```
# 
# ```{r}
# gastric_nsaf_missing <- gastric_nsaf_gene_level[rownames(gastric_nsaf_gene_level) %in% missing_proteins, ]
# 
# N <- 1
# keep <- rowSums(gastric_nsaf_missing > 0, na.rm = T) >= N
# missing <- rownames(gastric_nsaf_missing[keep,])
# ```


```{r}
mean(MPs$`Gastric Ascites` %in% detected_MP_proteins)
mean(detected_MP_proteins %in% MPs$`Gastric Ascites`)
```


```{r}
detected_MP_proteins[!(detected_MP_proteins %in% MPs$Gastric)]
```


# ```{r}
# x <- SpC_lists$Gastric$matrix
# x <- x[rowSums(x > 0) > 0, ]
# ```

```{r}
SpC_lists$Gastric$annotation
```


```{r}
colnames(gastric_nsaf) %in% rownames(SpC_lists$Gastric$annotation)
```

