---
title: "rna_seq_analysis"
output: html_document
date: "2022-09-27"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```



```{r}
library(tidyverse)
```


```{r}
count <- read_csv('data/merged_IPAS_RNAseq_counts_gastric.csv')
count <- count[-c(1:5), ]
colnames(count) <- c('Ensemble_ID', colnames(count)[-1])
count$Ensemble_ID <-str_extract(count$Ensemble_ID, pattern = '.+(?=\\.)')
count
```


## data wrangling
combine duplicate ensemble_id
63,920 -> 63,875
```{r}
count <- count %>% group_by(Ensemble_ID) %>% summarise_all(sum)
```


63,875 -> 43,939
remove rows that all cells are 0
```{r}
count <- count[-which(rowSums(count[-1]) == 0),]
```




import missing protein information
```{r}
## read-in newest missing protein list
missing_protein_df <- readxl::read_xlsx('data/PE2-5.xlsx')
missing_protein_df <- missing_protein_df[-c(1:12), ]
# head(missing_protein_df)
accession_to_symbol <- missing_protein_df %>% dplyr::select(Accession, `gene name(s)`) 
accession_to_symbol <- base::data.frame(gene_name = accession_to_symbol$`gene name(s)`, row.names = accession_to_symbol$Accession)
```



# Section 0: exploration

## overall distribution
```{r}
count_df <- as.data.frame(count[-1])
pseudo_counts <- log(count_df + 1)
ggplot(data = data.frame(x=c(as.matrix(pseudo_counts))), aes(x=x)) + 
  geom_histogram(aes(y = ..density..),
                 colour = 1, fill = "white") +
  geom_density(lwd = 1, colour = 4,
               fill = 4, alpha = 0.25)
```


## tag-wise distribution
```{r}
sample_filter <- sample(1:nrow(count_df), 100)
density_plot_dat <- t(count_df[sample_filter,]) %>% as_tibble() %>% pivot_longer(cols = everything(), names_to = 'ensemble_id', values_to = 'count') 
plot <- ggplot(data = density_plot_dat, aes(x=count, color=ensemble_id)) + 
  geom_density(alpha = 0.2) +
  scale_x_continuous(breaks=0:10, limits=c(0, 10)) +
  theme(legend.position="none")
suppressWarnings(print(plot))
```

## sample-wise distribution
```{r}
boxplot(log(count_df+1))
```




# Section 1 data processing


# do the conversion using biomaRt
the result is better than using org.Hs.eg.db
no missing gene symbol

! there are some Ensembl gene id cannot be found by biomart
https://support.bioconductor.org/p/111608/
```{r}
# mart <- biomaRt::useDataset("hsapiens_gene_ensembl", useMart("ensembl"))
# dict_ensembl_to_symbol <- biomaRt::getBM(filters= "ensembl_gene_id", attributes= c("ensembl_gene_id","hgnc_symbol", "start_position","end_position"), values=count$Ensemble_ID, mart= mart)
# dict_ensembl_to_symbol$gene_length=dict_ensembl_to_symbol$end_position - dict_ensembl_to_symbol$start_position
# dict_ensembl_to_symbol[dict_ensembl_to_symbol == ''] <- NA
# save(dict_ensembl_to_symbol, file='cache/dict_ensemble_to_symbol.rda')
load('cache/dict_ensemble_to_symbol.rda')
dict_ensembl_to_symbol

# code below can be used to check those unmapped ensembl gene id
# count$Ensemble_ID[!count$Ensemble_ID %in% dict_ensembl_to_symbol$ensembl_gene_id]
```





## filtering unmapped ensemble gene id
```{r}
count <- count %>% filter(Ensemble_ID %in% dict_ensembl_to_symbol$ensembl_gene_id)
```



## recreate data.frame and matrix
after several necessary pre-processing create data.frame and matrix version of the data
```{r}
count_df <- as.data.frame(count[-1])
rownames(count_df) <- count$Ensemble_ID
count_mat <- as.matrix(count_df)
```



normalization: calculate  TPM (transcripts per million)
```{r}
gene_length_df <- dict_ensembl_to_symbol %>% dplyr::select(ensembl_gene_id, gene_length) %>% unique()
gene_length_df <- data.frame(length=gene_length_df$gene_length, row.names=gene_length_df$ensembl_gene_id)
count_TPM_mat <- TPM(count_mat, gene_length_df)
head(count_TPM_mat)
save(count_TPM_mat, file='cache/count_TPM.rda')
```


Don't use TPM for differentially expression analysis: https://support.bioconductor.org/p/98820/#98875
The cutoff using raw count > 5 and TPM is actually similar: 0.848% of keep of remove decision are same
```{r}
keep <- rowSums(count_TPM_mat > 1, na.rm = T) >= ncol(count_TPM_mat)/2
keep2 <-  rowSums(count_mat > 5, na.rm = T) >= ncol(count_mat)/2
count_mat_filtered <- count_mat[keep,]
# number of row before filtering 63920
nrow(count_mat)
# number of row after filtering 7363
nrow(count_mat_filtered)
```

<!-- ```{r} -->
<!-- library(org.Hs.eg.db) -->
<!-- dict <- AnnotationDbi::select(org.Hs.eg.db, keys = rownames(count_mat), keytype='ENSEMBL', columns = c('SYMBOL', 'ENSEMBL')) -->
<!-- ``` -->


get the gene symbol in the gastric RNA-seq data and for those considered as transcripted genes
```{r}
all_rna_symbol_in_data <- dict_ensembl_to_symbol %>% filter(ensembl_gene_id %in% rownames(count_mat)) %>% filter(!is.na(hgnc_symbol)) %>% dplyr::select(hgnc_symbol) %>% pull()

expressed_rna_symbol <- dict_ensembl_to_symbol %>% filter(ensembl_gene_id %in% rownames(count_mat_filtered)) %>% filter(!is.na(hgnc_symbol)) %>% dplyr::select(hgnc_symbol) %>% pull()
# missing proteins that has transcripts level evidence in the data
RNA_supported_MPs <- expressed_rna_symbol[expressed_rna_symbol %in% missing_protein_df$`gene name(s)`] 
```



Get the list of expressed MPs in gene symbol
```{r}
convert_accession_to_symbol <- function(protein_accessions) {
  return(accession_to_symbol[protein_accessions, ])
}


# 
MPs_gene_symbols <- lapply(MPs, convert_accession_to_symbol)
```


```{r}
MP_gene_symbols <- MPs_gene_symbols$TCE

table(missing_protein_df %>% filter(`gene name(s)` %in% MP_gene_symbols & !is.na(`gene name(s)`)) %>% dplyr::select(PE) %>% pull)

MPs_has_trans_level_evidence <-  missing_protein_df %>% filter(`gene name(s)` %in% MP_gene_symbols & !is.na(`gene name(s)`)) %>% 
  filter(PE=='Evidence at transcript level') %>% dplyr::select(`gene name(s)`) %>% pull()

mean(MPs_has_trans_level_evidence %in% RNA_supported_MPs)
```



# !need also check just in gastric cancer TCE
```{r}
mean(gene_names %in% RNA_supported_MPs)
gene_names[gene_names %in% RNA_supported_MPs]
```



# check how many missing proteins extract the data is in the rna data
# make sure the conversion is compatible!!!!!


```{r}
lapply(MPs_gene_symbols, function(x) {mean(x %in% all_rna_symbol_in_data)})
```

```{r}
lapply(MPs_gene_symbols, function(x) {mean(x %in% expressed_rna_symbol)})
```


lets see how many MPs doesn't supported by RNA
```{r}
SpC_list_gastric <- SpC_List(SpC_TCE, annotation = annotations$TCE, NA_substitution = 0)

no_gastric_cancer <- unique(annotations$TCE$disease)[unique(annotations$TCE$disease) != 'Gastric']
replicates_for_removing_TCE <-  extract_sample_id(SpC_list_gastric, no_gastric_cancer)
replicates_for_removing_TCE <- c(replicates_for_removing_TCE, 'Prostate.IP0872_1701.H660')

SpC_list_gastric <- SpC_List(SpC_TCE, annotation = annotations$TCE, NA_substitution = 0, proteins_filter = missing_protein_df$Accession, replicates_remove = replicates_for_removing_TCE)

```

```{r}
MPs_gastric <- prefilter(SpC_list_gastric)
```

```{r}
MPs_gastric_gene_symbol <- convert_accession_to_symbol(MPs_gastric)
```

```{r}
mean(MPs_gastric_gene_symbol %in% all_rna_symbol_in_data)
```

```{r}
library(ggVennDiagram)
x <- list(A=MPs_gastric_gene_symbol, B=all_rna_symbol_in_data)
ggVennDiagram(x)
```


```{r}
mean(MPs_gastric_gene_symbol %in% expressed_rna_symbol)
```

```{r}
x <- list(A=MPs_gastric_gene_symbol, B=expressed_rna_symbol)
ggVennDiagram(x)
```

```{r}
MPs_has_trans_level_evidence <-  missing_protein_df %>% filter(`gene name(s)` %in% MPs_gastric_gene_symbol & !is.na(`gene name(s)`)) %>% 
  filter(PE=='Uncertain') %>% dplyr::select(`gene name(s)`) %>% pull()
```


# calculate correlation 